### DISCUSSION TITLE
코드의 출력 결과와 이유를 설명해주세요.

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#53] [스코프]
- 코드의 출력 결과와 이유를 설명해주세요.
   ```js
    var x = 1;
    
    function foo() {
	    var x = 10;
	    bar();
    }
    
    function bar() {
	    console.log(x);
    }
    
    foo();
    bar();
    ```
  - 렉시컬 스코프가 무엇인가요?
  - 동적 스코프와 렉시컬 스코프의 차이점이 무엇인가요?

**Comment by publdaze** [comment]
> - 코드의 출력 결과와 이유를 설명해주세요.
>    ```js
>     var x = 1;
>     
>     function foo() {
> 	    var x = 10;
> 	    bar();
>     }
>     
>     function bar() {
> 	    console.log(x);
>     }
>     
>     foo();
>     bar();
>     ```

- 함수가 호출되는 부분을 보면 먼저 foo 함수가 호출됨
- foo 함수 내부에서 bar 함수가 호출됨
- bar 함수 내부에서 x를 출력하고 있음
- foo 함수 내부에서 bar을 호출하였지만 자바스크립트에서는 렉시컬 스코프를 따르기 때문에 호출된 위치가 아닌 정의된 위치 기준으로 스코프가 결정됨
- 따라서 x 변수를 스코프 체인에 따라 변수를 찾아갈 때 foo 함수가 상위 스코프가 아닌 전역 스코프가 상위 스코프가 되고 최종적인 출력이 1이됨.
- 가장 하단 bar함수의 호출에서도 마찬가지고 x의 상위 스코프인 전역 스코프의 x 변수는 1이라는 값을 가지고 있기 때문에 1이 출력되게 됨

>  - 렉시컬 스코프가 무엇인가요?

- 렉시컬 스코프는 함수가 정의된 위치 기준으로 상위 스코프가 결정되는 것

>  - 동적 스코프와 렉시컬 스코프의 차이점이 무엇인가요?

- 동적 스코프는 함수가 호출된 위치 기준으로 상위 스코프가 결정되는 것
- 렉시컬 스코프는 정적 스코프라고도 하며 함수가 정의가 평가되는 시점에 정적으로 상위 스코프가 결정됨

**Comment by pipisebastian** [comment]
> 코드의 출력 결과와 이유를 설명해주세요

- 1,1
- 먼저 foo 함수에서 bar 를 실행했을때, 참조할 x를 찾기 위해 상위스코프로 이동합니다. 상위 스코프는 전역 스코프이며, 여기서 x 는 1 이기에 1을 출력합니다
- 다음으로 전역 스코프에서 호출한 bar함수의 경우에도, 앞과 똑같이 1을 출력합니다.
- 이러한 결과가 나오는 이유는 자바스크립트는 기본적으로 함수가 정의된 위치를 기준으로 스코프가 결정되기 때문입니다. 이러한 방식을 렉시컬 스코프라고 합니다.

> 렉시컬 스코프가 무엇인가요? + 동적 스코프와 렉시컬 스코프의 차이점이 무엇인가요?

- 스코프를 결정하는 방식인데요, 동적스코프는 함수를 어디서 호출했냐에따라 상위스코프가 결정되나, 렉시컬스코프는 함수가 정의된 위치에따라 상위스코프가 결정됩니다.
- 자바스크립트는 렉시컬 스코프를 따릅니다.
